### ISSUE TITLE
[Pydantic V2] Type checking in validators fails when reserializing models across threads

### ISSUE THREAD

**Issue by blipk** [pydantic/pydantic#7822]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I'm serializing models to disk using msgpack and getting the following contradictory error:

```
10 validation errors for RoleGroup
tasks.0
  Input should be a valid dictionary or instance of RoleTask [type=model_type, input_value=RoleTask(role_name='Comme...s: 'joined_paragraphs'>), input_type=RoleTask]
    For further information visit https://errors.pydantic.dev/2.4/v/model_type
tasks.1
  Input should be a valid dictionary or instance of RoleTask [type=model_type, input_value=RoleTask(role_name='Comme...s: 'joined_paragraphs'>), input_type=RoleTask]
    For further information visit https://errors.pydantic.dev/2.4/v/model_type
```

This is the models:
```Python
class RoleTask(BaseModel):
    role_name: str
    task_name: str

class Role(BaseModel):
    role_name: str

class RoleGroup(Role):
    tasks: list[RoleTask]
```

This is how it's serialized and deserialized
```Python

def decoder(obj: T | dict, force_type: str | None = None) -> T:
    if "__serialized_type__" in obj:
        obj_type = obj["__serialized_type__"]

        if obj_type == "RoleGroup" or force_type == "RoleGroup":
            obj = RoleGroup(**obj["class_dict"])
        elif obj_type == "RoleTask" or force_type == "RoleTask":
            obj = RoleTask(**obj["class_dict"])
    return obj
       
def type_check(
    obj: T, class_ref: type[T], class_str: str, force_type: bool = None
) -> bool:
    """
    isinstance and is class are failing across threads when they shouldn't.
    Possibly to do with PYTHONHASHSEED or import hierarchy.

    obj.__class__.__name__ seems to work best
    """
    try:
        is_instance = isinstance(obj, class_ref)
    except TypeError:
        is_instance = False

    return (
        is_instance
        or type(obj) is class_ref
        or (hasattr(obj, "__class__") and obj.__class__ is class_ref)
        or (
            hasattr(obj, "__class__")
            and hasattr(obj.__class__, "__name__")
            and obj.__class__.__name__ == class_str
        )
        or force_type == class_str
    )


def encoder(obj: T, force_type: str | None = None) -> dict | T:
    serialized = None

    if type_check(
        obj, RoleGroup, "RoleGroup", force_type
    ):
        obj: RoleGroup
        attr_dict = obj.__dict__.copy()
        # ERRORS HERE when constructing new_class
        new_class = RoleGroup(**attr_dict)
        serialized = {
            "__serialized_type__": "RoleGroup",
            "class_dict": new_class.__dict__,
        }
    elif type_check(obj, RoleTask, "RoleTask", force_type):
        obj: RoleTask
        attr_dict = obj.__dict__.copy()
        new_class = RoleTask(**attr_dict) 
        serialized = {
            "__serialized_type__": "RoleTask",
            "class_dict": new_class.__dict__,
        }

    return serialized if serialized else obj

task = RoleTask(role_name="test", task_name="task")
my_role_group = RoleGroup(role_name="role", tasks=[task])

import msgpack
binary_encoded: bytes = msgpack.packb(my_role_group, default=encoder)

reseriazlied_role_group: RoleGroup = msgpack.unpackb(binary_encoded, object_hook=decoder)
```

Note it only mostly fails when it is serialized in one thread, and then reserialized in another.

I had a similar problem with isinstance() builtin returning `False` even though it was showing it as an instance when I printed it's type(), not sure if it's related though, as it's failing in the encoder.

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
        pydantic-core version: 2.10.1
          pydantic-core build: profile=release pgo=true
                 install path: /home/user/.local/lib/python3.11/site-packages/pydantic
               python version: 3.11.5 (main, Sep  2 2023, 14:16:33) [GCC 13.2.1 20230801]
                     platform: Linux-6.4.16-5-MANJARO-x86_64-with-glibc2.38
             related packages: mypy-1.5.1 pydantic-settings-2.0.3 email-validator-2.0.0.post2 fastapi-0.88.0 typing_extensions-4.8.0
```
```

**Comment by dmontagu** [comment]
> More specifically, I think pydantic might be having a bit of trouble with loading data that has been deserialized from some non-JSON format.

To expand on this — we currently have different strictness rules when validating data directly from JSON vs. from data that has been loaded into python. I'm not sure if it's relevant here, but I'm thinking that _might_ be the source of the error you are hitting with the msgpack decoder.

If Sydney's suggestion to use `RoleGroup.model_validate_json(json.dumps(obj['class_dict']))` doesn't resolve the issue, it would help a lot if you could produce a self-contained example that we could run to hit the same error, which we could then use to try to better debug what's happening and how to resolve it. Even if it does resolve the issue, having the self-contained example would probably help us to work toward a solution that works for your use case without the extra serialize/deserialize pass.

**Comment by blipk** [comment]
Yeah I tried to create an isolated example, but it doesn't seem to happen there 🤷

It's got nothing to do with JSON, I'm using the __dict__ from my model to initialize a new model, it's actually in the encoder where it fails before it's even been serialized, but that is running in a different thread from where the original object was constructed - still strange if you look at the error it says it requires a `RoleTask` but I've provded a `RoleTask`.

I think it's because I'm writing this app in streamlit, and it has it's own script runner which wraps and execs() imports, I think they're doing something wrong there.

Anyway, I just removed that line and serialized the dict from the original object and it works fine, can't remember why I was reconstructing a new object but V1 required it for some reason.

I will close this as it's a strange issue and not really effecting me anymore.
